cpp/dolfinx/fem/dofmapbuilder.cpp

```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
+        }
       }
-      needs_entities[d] = true;
-      num_entities[d] = topology.index_map(d)->size_local()
-                        + topology.index_map(d)->num_ghosts();
-      local_size += num_entities[d] * element_dof_layout.num_entity_dofs(d);
     }
   }
 
-  // Collect cell -> entity connectivities
-  std::vector<std::shared_ptr<const graph::AdjacencyList<std::int32_t>>>
-      connectivity;
-  for (std::size_t d = 0; d < D; ++d)
-    connectivity.push_back(topology.connectivity(D, d));
-
-  // Allocate dofmap memory
-  const std::int32_t num_cells = topology.index_map(D)->size_local()
-                                 + topology.index_map(D)->num_ghosts();
-  dofmap_t dofs;
-  dofs.width = element_dof_layout.num_dofs();
-  dofs.array.resize(dofs.width * num_cells);
-
-  // Loop over cells and build dofmap
-  const std::vector<std::vector<std::vector<int>>> entity_dofs
-      = element_dof_layout.entity_dofs_all();
-  assert(entity_dofs.size() == D + 1);
-
-  for (std::int32_t c = 0; c < num_cells; ++c)
+#ifndef NDEBUG
   {
-    // Wrap dofs for cell c
-    std::span dofs_c(dofs.array.data() + c * dofs.width, dofs.width);
+    // Debug output
+    std::stringstream s;
+    s << "Required entities:";
+    for (std::size_t i = 0; i < required_dim_et.size(); ++i)
+      s << "(" << (int)required_dim_et[i].first << ", "
+        << (int)required_dim_et[i].second << ")=" << num_entity_dofs_et[i]
+        << " ";
+    LOG(INFO) << s.str();
+  }
+#endif
 
-    // Iterate over each topological dimension for this element
-    std::int32_t offset_local = 0;
-    for (std::size_t d = 0; d <= D; ++d)
+  // Dofmaps on each cell type as (width, [cell_dofs])
+  std::vector<dofmap_t> dofs(num_cell_types);
+
+  for (std::size_t i = 0; i < num_cell_types; ++i)
+  {
+    mesh::CellType cell_type = topology.entity_types(D)[i];
+    const std::vector<std::vector<std::vector<int>>> entity_dofs
+        = element_dof_layouts[i].entity_dofs_all();
+    assert(entity_dofs.size() == D + 1);
+
+    // Loop over cells of this type, and build dofmap
+    std::int32_t num_cells = topology.index_maps(D)[i]->size_local()
+                             + topology.index_maps(D)[i]->num_ghosts();
+    std::int32_t dofmap_width = element_dof_layouts[i].num_dofs();
+    dofs[i].width = dofmap_width;
+    dofs[i].array.resize(num_cells * dofmap_width);
+    LOG(INFO) << "Cell type:" << i << ", dofmap:" << num_cells << "x"
+              << dofmap_width;
+
+    std::int32_t dofmap_offset = 0;
+    for (std::int32_t c = 0; c < num_cells; ++c)
     {
-      if (needs_entities[d])
+      // Wrap dofs for cell c
+      std::span<std::int32_t> dofs_c(dofs[i].array.data() + dofmap_offset,
+                                     dofmap_width);
+      dofmap_offset += dofmap_width;
+
+      // Iterate over required entities for this element, dimension and type
+      for (std::size_t k = 0; k < required_dim_et.size(); ++k)
       {
+        std::size_t d = required_dim_et[k].first;
```

Comment:
`d`?

Reply:
Used throughout for topological dimension. I will add a comment here.